### ISSUE TITLE
Solution with classic csproj depending on one core csproj can not build if parallel builds are ON

### ISSUE THREAD

**Issue by davkean** [dotnet/sdk#1551]
_From @abpiskunov on August 29, 2017 17:2_

For consistent repro please use our solution $/DevDiv/Offcycle/WPT/WebToolsExtensions/Main/ProjectSystem.sln (you would need to sync whole Main folder and run tools\build.cmd from Dev Cmd line first. Then Open solution in VS and build it several times after some changes and without changes etc)

Scenario:

- Our solution contains ~15 projects including one core project ProjectSystem.Web
- some classic projects (unit tests) are referencing core project (this is just for the note, and i think is key to the repro)
- When we build our solution and have parallel builds have default value 4, we see random build failures, saying that there no access to some dlls dependencies being copied to the output folder.
- we have common output folder and before converting to core for that one project everything was working fine.

Note: I said random build errors, so some time there no errors and build is OK, but in some cases errors happening everytime and we just cannot build at all from VS, have to turn off parallel builds.

_Copied from original issue: dotnet/project-system#2762_

**Comment by davkean** [comment]
Can you provide build errors/targets in play so that I can route this to the correct team?

**Comment by davkean** [comment]
_From @abpiskunov on August 30, 2017 0:14_

Restoring NuGet packages...
To prevent NuGet from restoring packages during build, open the Visual Studio Options dialog, click on the Package Manager node and uncheck 'Allow NuGet to download missing packages during build.'
1>------ Build started: Project: Microsoft.VisualStudio.ProjectSystem.Web.15.0, Configuration: Debug Any CPU ------
2>------ Build started: Project: Microsoft.VisualStudio.Web.Publish.Contracts, Configuration: Debug Any CPU ------
3>------ Build started: Project: Microsoft.VisualStudio.Web.Common.TestParameters, Configuration: Debug Any CPU ------
1>C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(4222,5): error MSB3021: Unable to copy file "Z:\dd\WTE\Main\\references\Dev15\Microsoft.VisualStudio.Shell.15.0.dll" to "Z:\dd\WTE\Main\\bin\dev15\Debug\Microsoft.VisualStudio.Shell.15.0.dll". Access to the path 'Z:\dd\WTE\Main\\bin\dev15\Debug\Microsoft.VisualStudio.Shell.15.0.dll' is denied.
3>  Microsoft.VisualStudio.Web.Common.TestParameters -> Z:\dd\WTE\Main\bin\dev15\Test\Debug\Microsoft.VisualStudio.Web.Common.TestParameters.dll
2>C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(4222,5): error MSB3021: Unable to copy file "Z:\dd\WTE\Main\\references\Dev15\Microsoft.VisualStudio.ImageCatalog.dll" to "Z:\dd\WTE\Main\\bin\dev15\Debug\Microsoft.VisualStudio.ImageCatalog.dll". Access to the path 'Z:\dd\WTE\Main\\bin\dev15\Debug\Microsoft.VisualStudio.ImageCatalog.dll' is denied.
2>C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(4222,5): warning MSB3026: Could not copy "C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Build.Tasks.Core.dll" to "Z:\dd\WTE\Main\\bin\dev15\Debug\Microsoft.Build.Tasks.Core.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'Z:\dd\WTE\Main\\bin\dev15\Debug\Microsoft.Build.Tasks.Core.dll' because it is being used by another process.
4>------ Build started: Project: Microsoft.VisualStudio.ProjectSystem.Web, Configuration: Debug Any CPU ------
4>C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(1987,5): warning MSB3277: Found conflicts between different versions of the same dependent assembly that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
4>Microsoft.VisualStudio.ProjectSystem.Web -> Z:\dd\WTE\Main\bin\dev15\Debug\Microsoft.VisualStudio.ProjectSystem.Web.dll
4>Done building project "Microsoft.VisualStudio.ProjectSystem.Web.csproj".
========== Build: 2 succeeded, 2 failed, 15 up-to-date, 0 skipped ==========

Seems like files are locked by dotnet and other csproj projects can not copy them?

**Comment by davkean** [comment]
Can you see if two vstest.XXX.exe process are running and If so kill them and see if that resolves the issue?

**Comment by davkean** [comment]
_From @abpiskunov on August 30, 2017 0:27_

There were 2 (why 2?) vstest.xxx processes , killed them , it did not help. But(!) there were 4 msbuild processes, killing all of them did help and next build did not have locked files.

**Comment by davkean** [comment]
There's 1 for every core, so 4 is expected. This is probably a task leaking a handle - no idea on the underlying cause, i'll move to SDK.

**Comment by livarcocc** [comment]
@abpiskunov, is this still reproing for you? Does this repro when the build is executed through the command line?

**Comment by abpiskunov** [comment]
Yes, I still can repro it when set number of parallel builds to 4 (we had to switch to 1 by default on our dev machines). When I run msbuild /m:4 ProjectSystem.Web.sln from command line it does not fail.

**Comment by Barsonax** [comment]
You also get similar random errors when parallel build is enabled and you use webpublish:
```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VisualStudio\v16.0\Web\Deploy\Microsoft.Web.Publishing.Deploy.FileSystem.targets(96,5): error : Copying file obj\Release\Package\PackageTmp\bin\ICSharpCode.SharpZipLib.dll to D:\a\1\s\buildoutput\cmcd\bin\ICSharpCode.SharpZipLib.dll failed. The process cannot access the file 'D:\a\1\s\buildoutput\cmcd\bin\ICSharpCode.SharpZipLib.dll' because it is being used by another process. [D:\a\1\s\src\Foundation\Macaw\ScheduledAgent\code\Foundation.Macaw.ScheduledAgent.csproj]
```

I think its caused by project A still doing the copy and thus having a file lock at the same time when project B wants to copy the same file. Msbuild doesn't seem to be able to handle this properly.

The issue is very intermittent and hard to reproduce but I seen it happen on multiple machines and also in azure pipelines.

**Comment by Barsonax** [comment]
Seems to be caused by the virus scanner. Adding a exclusion to your source folder will prevent these errors it seems.
